workflow/workflow.go

```diff
@@ -265,65 +265,87 @@ func GetSignalChannel(ctx Context, signalName string) Channel {
 	return internal.GetSignalChannel(ctx, signalName)
 }
 
-// SideEffect executes the provided function once, records its result into the workflow history. The recorded result on
-// history will be returned without executing the provided function during replay. This guarantees the deterministic
-// requirement for workflow as the exact same result will be returned in replay.
-// Common use case is to run some short non-deterministic code in workflow, like getting random number or new UUID.
-// The only way to fail SideEffect is to panic which causes decision task failure. The decision task after timeout is
-// rescheduled and re-executed giving SideEffect another chance to succeed.
-//
-// Caution: do not use SideEffect to modify closures. Always retrieve result from SideEffect's encoded return value.
-// For example this code is BROKEN:
+// SideEffect executes the provided callback, and records its result into the workflow history.
+// During replay the recorded result will be returned instead, so the callback can do non-deterministic
+// things (such as reading files or getting random numbers) without breaking determinism during replay.
+//
+// As there is no error return, the callback's code is assumed to be reliable.
+// If you cannot retrieve the value for some reason, panicking and failing the decision task
+// will cause it to be retried, possibly succeeding on another machine.
+// For better error handling, use ExecuteLocalActivity instead.
+//
+// Caution: the callback MUST NOT call any blocking or history-creating APIs,
+// e.g. workflow.Sleep or ExecuteActivity or calling .Get on any future.
+// This will (potentially) work during the initial recording of history, but will
+// fail when replaying because the data is not available when the call occurs
+// (it becomes available later).
+//
+// In other words, in general: use this *only* for code that does not need the workflow.Context.
+// Incorrect context-using calls are not currently prevented in tests or during execution,
+// but they should be eventually.
+//
+//	// Bad example: this will work until a replay occurs,
+//	// but then the workflow will fail to replay with a non-deterministic error.
+//	var out string
+//	err := workflow.SideEffect(func(ctx workflow.Context) interface{} {
```

Comment:
Is there a ticket/initiative to fix this?
We cannot change the API since it is used everywhere, but we could introduce workflow.Context check will fail in the case of history-related operations inside these calls.
Something like:
workflow.Context{
  insideSidefffect: true
}

and inside forbidden things bail out if this flag is set.